File: backend/src/main/java/com/culinaryheaven/domain/recipe/dto/response/RecipeResponse.java
package com.culinaryheaven.domain.recipe.dto.response;

import com.culinaryheaven.domain.contest.dto.response.ContestResponse;
import com.culinaryheaven.domain.recipe.domain.Recipe;
import io.swagger.v3.oas.annotations.media.Schema;

import java.util.List;

public record RecipeResponse(

        @Schema(description = "레시피 ID")
        Long id,

        @Schema(description = "레시피 제목")
        String title,

        @Schema(description = "레시피 설명")
        String description,

        @Schema(description = "레시피 대표 이미지 URL")
        String thumbnailImageUrl,

        @Schema(description = "소유자 정보")
        WriterInfoResponse writerInfoResponse,

        @Schema(description = "레시피 스텝")
        List<StepResponse> steps,

        @Schema(description = "사용된 재료")
        List<IngredientResponse> ingredients,

        @Schema(description = "레시피가 등록된 대회 정보")
        ContestResponse contest,

        @Schema(description = "북마크 개수")
        int bookMarkCounts,

        @Schema(description = "좋아요 개수")
        int likeCounts,

        @Schema(description = "북마크 여부")
        Boolean isBookMarked,

        @Schema(description = "좋아요 여부")
        Boolean isLiked,

        @Schema(description = "소유자 여부")
        Boolean isOwner
) {
    public static RecipeResponse of(Recipe recipe, Boolean isBookMarked, Boolean isLiked, Boolean isOwner) {
        return new RecipeResponse(
                recipe.getId(),
                recipe.getTitle(),
                recipe.getDescription(),
                recipe.getThumbnailImage(),
                WriterInfoResponse.of(recipe.getUser()),
                StepsResponse.of(recipe.getSteps()).steps(),
                IngredientsResponse.of(recipe.getIngredients()).ingredients(),
                ContestResponse.of(recipe.getContest()),
                recipe.getBookMarkCount(),
                recipe.getLikeCount(),
                isBookMarked,
                isLiked,
                isOwner
        );
    }
}

File: android/data/src/main/java/com/jae464/data/remote/model/response/recipe/RecipeResponse.kt
package com.jae464.data.remote.model.response.recipe

import com.jae464.data.remote.model.response.contest.ContestResponse
import com.jae464.data.util.adjustLocalhostUrl
import com.jae464.domain.model.Ingredient
import com.jae464.domain.model.Recipe
import com.jae464.domain.model.Step
import kotlinx.serialization.Serializable

@Serializable
data class RecipeResponse(
    val id: Long,
    val title: String,
    val description: String,
    val thumbnailImageUrl: String,
    val writerInfoResponse: WriterInfoResponse,
    val steps: List<StepResponse>,
    val ingredients: List<IngredientResponse>,
    val contest: ContestResponse,
    val bookMarkCounts: Int,
    val likeCounts: Int,
    val isBookMarked: Boolean,
    val isLiked: Boolean,
    val isOwner: Boolean?
)

@Serializable
data class StepResponse(
    val id: Long,
    val step: Int,
    val description: String,
    val imageUrl: String?
)

@Serializable
data class IngredientResponse(
    val id: Long,
    val name: String,
    val quantity: String
)

fun RecipeResponse.toDomain() = Recipe(
    id = id,
    title = title,
    imageUrl = adjustLocalhostUrl(thumbnailImageUrl),
    description = description,
    score = 5f,
    writeInfo = writerInfoResponse.toDomain(),
    ingredients = ingredients.map { it.toDomain() },
    steps = steps.map { it.toDomain() },
    bookMarkCounts = bookMarkCounts,
    likeCounts = likeCounts,
    isLiked = isLiked,
    isBookMarked = isBookMarked,
    isOwner = isOwner ?: false
)

fun IngredientResponse.toDomain() = Ingredient(
    name = name,
    quantity = quantity
)

fun StepResponse.toDomain() = Step(
    step = step,
    description = description,
    imageUrl = imageUrl
)
